cpp/src/arrow/compute/kernels/scalar_set_lookup.cc

```diff
@@ -75,6 +95,9 @@ struct SetLookupState : public KernelState {
 
   using MemoTable = typename HashTraits<Type>::MemoTableType;
   MemoTable lookup_table;
+  // When there are duplicates in value_set, the MemoTable indices must
+  // be mapped back to indices in the value_set.
+  std::vector<int32_t> memo_index_to_value_index;
```

Comment:
If there are not many duplicated values (common case?), looks it's possible to compress this table by storing only indices where duplicated values are met.

A quick example. I think we can leave it as follow up task if it works.
```
values       a    a    b    a    b    b    c

value_index  0    1    2    3    4    5    6

// memo index to value index mapping
memo_index_to_value_index[0] = 0  --> a
memo_index_to_value_index[1] = 2  --> b
memo_index_to_value_index[2] = 6  --> c

// duplicated value indices in sorted order
duplicated_indices = (1, 3, 4, 5)

// lookup 'a'
memo_index = lookup('a') = 0
0 < duplicated_indices[0]
return 0

// lookup 'b'
memo_index = lookup('b') = 1
1 >= duplicated_indices[0]
memo_index++   // = 2
2 < duplicated_indices[1]
return 2

// lookup 'c'
memo_index = lookup('c') = 2
2 >= duplicated_indices[0]
memo_index++   // = 3
3 >= duplicated_indices[1]
memo_index++   // = 4
4 >= duplicated_indices[2]
memo_index++   // = 5
5 >= duplicated_indices[3]
memo_index++   // = 6
return 6
```

Reply:
We would some lose performance in exchange of a little space saving. The hash table is much larger, I don't think this is desirable.